src/chain/chain-supervision.cc

```diff
@@ -207,6 +209,47 @@ bool PhoneLatticeToProtoSupervision(const SupervisionOptions &opts,
   return true;
 }
 
+bool NormalizePhoneLattice(CompactLattice *clat) {
+  if (clat->Properties(fst::kTopSorted, false) == 0) {
+    if (fst::TopSort(clat) == false) {
+      KALDI_WARN << "Cycles detected in lattice: cannot normalize.";
+      return false;
+    }
+  }
+
+  std::vector<double> beta;
+  if (!ComputeCompactLatticeBetas(*clat, &beta)) {
+    KALDI_WARN << "Failed to compute backward probabilities on lattice.";
+    return false;
+  }
+
+  CompactLattice::Arc::StateId start = clat->Start();  // Should be 0
+  BaseFloat total_backward_cost = beta[start];
+
+  for (fst::StateIterator<CompactLattice> sit(*clat); !sit.Done(); sit.Next()) {
+    CompactLatticeWeight f = clat->Final(sit.Value());
+    LatticeWeight w = f.Weight();
+    w.SetValue1(w.Value1() + total_backward_cost);
```

Comment:
Actually I don't believe this code is quite right for this purpose..  The total_backward_cost can be interpreted as the negative of value1() + value2() of the total weight, and you are just adding it to the value1().  It's the value1() which we care about being normalized.  But it won't end up normalized if you process it like this, it will have something with the magnitude of the value2() added to it.   It would be OK if you were to zero out all the value2()'s first.  But I think it would be easier if you were to do the normalization in a different way-- after converting it to an FST.  You could compute the best-path cost using ShortestDistance, and then subtract that from all the final-probs.
This may require a slight refactoring of the code.